cmd/cluster/core/destroy.go

```diff
@@ -90,67 +90,65 @@ func GetCluster(ctx context.Context, o *DestroyOptions) (*hyperv1.HostedCluster,
-		if hostedCluster.DeletionTimestamp == nil {
-			controllerutil.AddFinalizer(hostedCluster, destroyFinalizer)
-		}
-		if o.DestroyCloudResources {
-			if hostedCluster.Annotations == nil {
-				hostedCluster.Annotations = map[string]string{}
+	if hostedClusterExists {
+		if shouldDestroyPlatformSpecifics {
+			err = setFinalizer(ctx, hostedCluster, o, c)
+			if err != nil {
+				return err
 			}
-			hostedCluster.Annotations[hyperv1.CleanupCloudResourcesAnnotation] = "true"
-		}
-		if err := c.Patch(ctx, hostedCluster, client.MergeFrom(original)); err != nil {
-			if apierrors.IsNotFound(err) {
-				o.Log.Info("Hosted cluster not found, skipping finalizer update", "namespace", o.Namespace, "name", o.Name)
-			} else if !strings.Contains(err.Error(), "no new finalizers can be added if the object is being deleted") {
-				return fmt.Errorf("failed to add finalizer to hosted cluster: %w", err)
-			}
-		} else {
-			o.Log.Info("Updated finalizer for hosted cluster", "namespace", o.Namespace, "name", o.Name)
 		}
+
 		o.Log.Info("Deleting hosted cluster", "namespace", o.Namespace, "name", o.Name)
-		if err := c.Delete(ctx, hostedCluster); err != nil {
+		if err = c.Delete(ctx, hostedCluster); err != nil {
 			if apierrors.IsNotFound(err) {
 				o.Log.Info("Hosted not found, skipping delete", "namespace", o.Namespace, "name", o.Name)
 			} else {
 				return fmt.Errorf("failed to delete hostedcluster: %w", err)
 			}
 		}
-		// Wait for the hosted cluster to have only the CLI's finalizer remaining,
-		// which should indicate the cluster was successfully torn down.
-		clusterDeleteCtx, clusterDeleteCtxCancel := context.WithTimeout(ctx, o.ClusterGracePeriod)
-		defer clusterDeleteCtxCancel()
-		err := wait.PollImmediateUntil(1*time.Second, func() (bool, error) {
-			if err := c.Get(clusterDeleteCtx, types.NamespacedName{Namespace: o.Namespace, Name: o.Name}, hostedCluster); err != nil {
-				if apierrors.IsNotFound(err) {
-					return true, nil
-				}
-				o.Log.Error(err, "Failed to get hosted cluster", "namespace", o.Namespace, "name", o.Name)
-				return false, nil
+
+		if shouldDestroyPlatformSpecifics {
+			if err = waitForRestOfFinalizers(ctx, hostedCluster, o, c); err != nil {
+				return err
 			}
-			done := len(hostedCluster.Finalizers) == 1 && hostedCluster.Finalizers[0] == destroyFinalizer
-			return done, nil
-		}, clusterDeleteCtx.Done())
-		if err != nil {
-			return fmt.Errorf("hostedcluster wasn't finalized, aborting delete: %w", err)
 		}
 	}
 
-	// Destroy additional resources which are specific to the current platform
-	if err := destroyPlatformSpecifics(ctx, o); err != nil {
-		return err
+	if shouldDestroyPlatformSpecifics {
+		// Destroy additional resources which are specific to the current platform
+		if err = destroyPlatformSpecifics(ctx, o); err != nil {
+			return err
+		}
 	}
 
 	// clean up CLI generated secrets
+	if err = deleteCLISecrets(ctx, o, c); err != nil {
+		return err
+	}
+
```

Comment:
deleting the secrets could have unintended consequences if it's performed  before the hosted cluster is either gone or finished tearing down.